### PR TITLE
[WPE] Add WPE context menu support

### PR DESCRIPTION
**Pull request by chrisduerr**
<pre>
Add WPE context menu support
<a href="https://bugs.webkit.org/enter_bug.cgi">https://bugs.webkit.org/show_bug.cgi?id=272354</a>

Reviewed by NOBODY (OOPS!).

This replaces the existing context menu stub implementation with a minimal version which emits the WebView's context menu signal.

This also resolves a bug where all mouse input would break after right-clicking, since the stub implementation was waiting for the context menu to be opened before processing new mouse events.

The `context-menu-dismissed` signal is removed from the WPE backend in a breaking change, since dismissal is handled by the consumer of the API.

* Source/WebKit/Shared/API/glib/WebKitContextMenuItem.cpp: (webkit_context_menu_item_get_title):
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitContextMenuItem.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp: (webkit_web_view_class_init):
* Source/WebKit/UIProcess/wpe/WebContextMenuProxyWPE.cpp: Copied from Source/WebKit/UIProcess/wpe/WebContextMenuProxyWPE.h. (WebKit::WebContextMenuProxyWPE::WebContextMenuProxyWPE): (WebKit::WebContextMenuProxyWPE::show):
(WebKit::WebContextMenuProxyWPE::showContextMenuWithItems):
* Source/WebKit/UIProcess/wpe/WebContextMenuProxyWPE.h:

</pre>

https://github.com/WebKit/WebKit/commit/dcde3dc0734583304c5a65a972b7dde90e539b75

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82140 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1687 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36115 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86738 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33178 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1729 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9517 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64019 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21752 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85210 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1275 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74737 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44300 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1174 "Found 4 new test failures: imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml.html imported/w3c/web-platform-tests/css/css-text/text-autospace/text-autospace-001.html imported/w3c/web-platform-tests/css/css-text/text-autospace/text-autospace-break-001.html imported/w3c/web-platform-tests/html/semantics/disabled-elements/event-propagate-disabled-keyboard.tentative.html (failure)") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31645 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72471 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29532 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88174 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9393 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6702 "Found 2 new test failures: http/tests/security/contentSecurityPolicy/connect-src-star-websocket-allowed.html pdf/pdf-plugin-printing.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72403 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9582 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70555 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71623 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15750 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14661 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/799 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9349 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14875 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9194 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12711 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10996 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/Shared/API/glib/WebKitContextMenuItem.cpp [hunk 1]
```diff
@@ -333,6 +333,28 @@ GAction* webkit_context_menu_item_get_gaction(WebKitContextMenuItem* item)
...
+ * Since: 2.18
+ */
+const char* webkit_context_menu_item_get_title(WebKitContextMenuItem* item)
+{
+    g_return_val_if_fail(WEBKIT_IS_CONTEXT_MENU_ITEM(item), nullptr);
+
+#if ENABLE(CONTEXT_MENUS)
+    return item->priv->menuItem->title().utf8().data();
+#else
+    g_assert_not_reached();
```
Comment (aperezdc) [inline Source/WebKit/Shared/API/glib/WebKitContextMenuItem.cpp]: GLib assertions may be removed at build time, and in that case we would need to return something here. I think `NULL` should be safe:

```suggestion
    g_assert_not_reached();
    return nullptr;
```

File: Source/WebKit/Shared/API/glib/WebKitContextMenuItem.cpp [hunk 2]
```diff
@@ -333,6 +333,28 @@ GAction* webkit_context_menu_item_get_gaction(WebKitContextMenuItem* item)
...
+/**
+ * webkit_context_menu_item_get_title:
+ * @item: a #WebKitContextMenuItem
+ *
+ * Gets the title associated to @item.
+ *
+ * Returns: (transfer none): the title associated to the #WebKitContextMenuItem,
+ *    or %NULL if @item is a separator.
+ *
+ * Since: 2.18
```
Comment (aperezdc) [inline Source/WebKit/Shared/API/glib/WebKitContextMenuItem.cpp]: The next stable series will be `2.48`, I guess the `1` there is a typo:

```suggestion
 * Since: 2.48
```

File: Source/WebKit/UIProcess/wpe/WebContextMenuProxyWPE.cpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+{
+}
+
+void WebContextMenuProxyWPE::show()
+{
+    WebContextMenuProxy::show();
+}
+
+void WebContextMenuProxyWPE::showContextMenuWithItems(Vector<Ref<WebContextMenuItem>>&& items)
+{
```
Comment (aperezdc) [inline Source/WebKit/UIProcess/wpe/WebContextMenuProxyWPE.cpp:45]: I think we could add a `notImplemented()` call here:

```suggestion
{
    notImplemented();
```

File: Source/WebKit/UIProcess/wpe/WebContextMenuProxyWPE.cpp [hunk 4]
```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (C) 2011, 2020 Igalia S.L.
```
Comment (aperezdc) [inline Source/WebKit/UIProcess/wpe/WebContextMenuProxyWPE.cpp]: You may want to put your name here instead, or the name of your employer 😉

**Review by aperezdc — CHANGES_REQUESTED** [review]
Patch LGTM; but there are a few small wrinkles.

@chrisduerr: Could you take a look and update the patch? Thanks for your time!

File: Source/WebKit/Shared/API/glib/WebKitContextMenuItem.cpp [hunk 5]
```diff
@@ -333,6 +333,28 @@ GAction* webkit_context_menu_item_get_gaction(WebKitContextMenuItem* item)
...
+ */
+const char* webkit_context_menu_item_get_title(WebKitContextMenuItem* item)
+{
+    g_return_val_if_fail(WEBKIT_IS_CONTEXT_MENU_ITEM(item), nullptr);
+
+#if ENABLE(CONTEXT_MENUS)
+    return item->priv->menuItem->title().utf8().data();
```
Comment (carlosgcampos) [inline Source/WebKit/Shared/API/glib/WebKitContextMenuItem.cpp]: This is returning a value that is released right after the function finishes, because String::utf8() returns a temporary. We need to cache the CString and returns its data to keep this returning a const char

File: Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp [hunk 6]
```diff
@@ -165,7 +165,9 @@ enum {
     RUN_FILE_CHOOSER,
 
     CONTEXT_MENU,
+#if PLATFORM(GTK)
     CONTEXT_MENU_DISMISSED,
+#endif
```
Comment (carlosgcampos) [inline Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp]: hmm, this an api break even when this never worked for wpe.

Comment (chrisduerr) [reply Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp]: I can add it back if that's simpler.

Comment (mcatanzaro) [reply Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp]: Yeah you can't ever remove signals from the public API. You could deprecate it for WPE, though. This would require moving the doc comment to a header file; unifdef will then respect `#if PLATFORM(GTK)` or `#if PLATFORM(WPE)` guards when generating the actual header that gets installed and scanned for doc comments.

Comment (mcatanzaro) [reply Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp]: You would add:

```
Deprecated: 2.48, WPE WebKit does not emit this signal
```

to the bottom of the doc comment, and then also change:

```
#if PLATFORM(GTK)
G_SIGNAL_RUN_LAST,
#else
static_cast<GSignalFlags>(G_SIGNAL_RUN_LAST | G_SIGNAL_DEPRECATED),
#endif
```

File: Source/WebKit/Shared/API/glib/WebKitContextMenuItem.cpp (hunk 5 above)
Comment (chrisduerr) [reply Source/WebKit/Shared/API/glib/WebKitContextMenuItem.cpp]: I've decided to just go with reassigning it every call. If a more complicated solution is desired that checks if the CString is still accurate to avoid re-conversion, please let me know.

**Comment by chrisduerr** [comment]
Based on other PRs, it seems like force-pushing is the standard practice for updates to WebKit PRs. Please let me know if this is not the case.

**Comment by mcatanzaro** [comment]
You have to force push, yes.

File: Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp (hunk 6 above)
Comment (chrisduerr) [reply Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp]: Thanks for laying it out so clearly for me, would have taken me much longer otherwise!

File: Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp [hunk 7]
```diff
@@ -2257,15 +2257,20 @@ static void webkit_web_view_class_init(WebKitWebViewClass* webViewClass)
...
-                     G_STRUCT_OFFSET(WebKitWebViewClass, context_menu_dismissed),
-                     0, 0,
-                     g_cclosure_marshal_VOID__VOID,
-                     G_TYPE_NONE, 0);
+    signals[CONTEXT_MENU_DISMISSED] = g_signal_new("context-menu-dismissed",
+        G_TYPE_FROM_CLASS(webViewClass),
+#if PLATFORM(GTK)
+        G_SIGNAL_RUN_LAST,
+#else
+        static_cast<GSignalFlags>(G_SIGNAL_RUN_LAST | G_SIGNAL_DEPRECATED),
```
Comment (chrisduerr) [inline Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp]: Formatter complained about the indentation of this new line despite (to my eyes) matching the existing indentation. So I've decided to just reformat the entire block to clangd's suggested format, hope that's good.

Comment (mcatanzaro) [reply Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp]: Yes, that was the right choice.

Comment (mcatanzaro) [reply Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp]: (Older class init functions often don't comply with our mandatory code style guidelines. We just reformat it when we touch the old code.)

**Comment by chrisduerr** [comment]
When implementing this with an additional touch support PR, I noticed that the arguments to the `connect_context_menu` signal handler include the web view, the context menu, and the hit test result. However none of these three objects give access to the actual *position* of the context menu's origin.

So I would hold off on this for now, I'll try and look into either getting the position out of a hit test result (I would have expected it to be stored on that), or pass it separately if neither context menu nor hit test result include it.

It's possible my bindings are just missing something, but I don't think that's the case here.

**Comment by chrisduerr** [comment]
I've looked at the context menu and hit test results and neither seem to have easy access to the original event location. So while it would probably be possible with a bunch of work to get that wired up, I feel inclined to suggest that users just keep track of the latest touch/click events and spawn the context menu at that position. That's definitely not super reliable, but I doubt it would have an impact on users in practice?

**Comment by aperezdc** [comment]
@carlosgcampos Any thoughts on API for exposing the location of the hit test results? Maybe it would be better to do that as a separate follow-up patch, what do you think?

**Comment by chrisduerr** [comment]
> I think this may be filled in with the coordinates of the mouse pointer for events triggered by the mouse. When that is not available, the WebHistTestResultData::elementBoundingBox could be a reasonable option to have an approximate location of where the focus is.

Do you know under which circumstances that might not be available? I don't think it's possible to open a context menu from JavaScript, so would this fallback only apply on hit test unrelated to the context menu?

I'm mainly asking because I don't think using the element itself as fallback for a context menu would be viable. If there's a scenario where that would pop up, it would likely be better to just do what I do right now and store the last input event position.

File: Source/WebKit/Shared/API/glib/WebKitContextMenuItem.cpp [hunk 8]
```diff
@@ -333,6 +334,30 @@ GAction* webkit_context_menu_item_get_gaction(WebKitContextMenuItem* item)
...
+ * @item: a #WebKitContextMenuItem
+ *
+ * Gets the title associated to @item.
+ *
+ * Returns: (transfer none): the title associated to the #WebKitContextMenuItem,
+ *    or %NULL if @item is a separator.
+ *
+ * Since: 2.48
+ */
+const char* webkit_context_menu_item_get_title(WebKitContextMenuItem* item)
```
Comment (carlosgcampos) [inline Source/WebKit/Shared/API/glib/WebKitContextMenuItem.cpp]: This is unrelated to WPE context menu support. I would move it to a separate patch and since it's new public API, it requires an API test.

File: Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp [hunk 9]
```diff
@@ -2257,15 +2257,20 @@ static void webkit_web_view_class_init(WebKitWebViewClass* webViewClass)
      *
      * Emitted after #WebKitWebView::context-menu signal, if the context menu is shown,
      * to notify that the context menu is dismissed.
+     *
+     * Deprecated: 2.48, WPE WebKit does not emit this signal.
```
Comment (carlosgcampos) [inline Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp]: This is deprecating the signal for GTK too. We need to split the documentation and definition like the context-menu signal.

**Review by carlosgcampos — CHANGES_REQUESTED** [review]
We need to add API tests to check the signal si emitted and we get the items to build a context menu.

**Comment by carlosgcampos** [comment]
> @carlosgcampos Any thoughts on API for exposing the location of the hit test results? Maybe it would be better to do that as a separate follow-up patch, what do you think?

The context menu works so differently between gtk and wpe, that maybe it would be better to deprecated current signals for wpe and a new one for wpe only where we can include the location as parameter, for example. In case of wpe we always expect the user to receive the items to build and show the menu, so we could add `WebKitWebView::show-context-menu`

**Comment by chrisduerr** [comment]
Unfortunately I have neither the time nor energy to finish upstreaming my WebKit patches and I think generally the interest of touch support is pretty low anyway. So I'll save you the reviews and just keep my stuff out of tree.
